src/js/DOMUtils.js

```diff
@@ -0,0 +1,199 @@
+ * @param {Node} node
+ * @return {?Node}
+ */
+axs.dom.composedParentNode = function(node) {
+    if (!node)
+        return null;
+    if (node.nodeType === Node.DOCUMENT_FRAGMENT_NODE)
+        return node.host;
+
+    var parentNode = node.parentNode;
+    if (!parentNode)
+        return null;
+
+    if (parentNode.nodeType === Node.DOCUMENT_FRAGMENT_NODE)
+        return parentNode.host;
+
+    var shadowRoot = parentNode.shadowRoot;
+    if (!parentNode.shadowRoot)
+        return parentNode;
+
+    var eventName = generateId();
+    var path;
+    parentNode.addEventListener(eventName, function listener(event) {
+       event.stopPropagation();
+       path = event.path;
+       parentNode.removeEventListener(eventName, listener);
+       return false;
+    })
+    var event = new CustomEvent(eventName, { bubbles: true });
+    node.dispatchEvent(event);
+    for (var i = 1; i < path.length; i++) {
+        var pathNode = path[i];
+        if (pathNode.localName == 'content')
+            continue;
+        if (pathNode.nodeType === Node.DOCUMENT_FRAGMENT_NODE)
+            continue;
+        return pathNode;
+    }
+    return null;
+};
+})();
+
+/**
+ * Return the corresponding element for the given node.
+ * @param {Node} node
+ * @return {Element}
+ * @suppress {checkTypes}
+ */
+axs.dom.asElement = function(node) {
+    /** @type {Element} */ var element;
+    switch (node.nodeType) {
+    case Node.COMMENT_NODE:
+        return null;  // Skip comments
+    case Node.ELEMENT_NODE:
+        element = /** (@type {Element}) */ node;
+        if (element.localName == 'script' ||
+            element.localName == 'template')
+            return null;  // Skip script-supporting elements
+        return element;;
+    case Node.DOCUMENT_FRAGMENT_NODE:
+        return node.host;
+    case Node.TEXT_NODE:
+        return axs.dom.parentElement(node);
+    default:
+        console.warn('Unhandled node type: ', node.nodeType);
+    }
+    return null;
+};
+
+/**
+ * Recursively walk the composed tree from |node|, aborting if |end| is encountered.
+ * @param {Node} node
+ * @param {?Node} end
+ * @param {{preorder: (function (Node)|undefined),
+ *          postorder: (function (Node)|undefined)}} callbacks
+ *     Callbacks to be called for each element traversed, excluding
+ *     |end|. Possible callbacks are |preorder|, called before descending into
+ *     child nodes, and |postorder| called after all child nodes have been
+ *     traversed.
+ * @param {ShadowRoot=} opt_shadowRoot The nearest ShadowRoot ancestor, if any.
+ * @return {boolean} Whether |end| was found, if provided.
+ */
+axs.dom.composedTreeSearch = function(node, end, callbacks, opt_shadowRoot) {
+    if (node === end)
+        return true;
+
+    if (node.nodeType == Node.ELEMENT_NODE)
+        var element = /** @type {Element} */ (node);
+
+    if (element && callbacks.preorder)
+        callbacks.preorder.call(null, element);
```

Comment:
It's not immediately obvious why it's `callbacks.preorder.call(null, element);` instead of `callbacks.preorder(element)`

Same below with `callbacks.postorder.call(null, element);`